federation/pkg/dnsprovider/providers/google/clouddns/clouddns_test.go

```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		return nil, err
+	}
+	return i, nil
+}
+
+func newFakeInterface() (dnsprovider.Interface, error) {
+	service := stubs.NewService()
+	interface_ := newInterfaceWithStub("", service)
+	zones := service.ManagedZones_
+	// Add a fake zone to test against.
+	zone := &stubs.ManagedZone{zones, "example.com", []stubs.ResourceRecordSet{}}
+	call := zones.Create(interface_.project(), zone)
+	_, err := call.Do()
+	if err != nil {
+		return nil, err
+	}
+	return interface_, nil
+}
+
+var interface_ dnsprovider.Interface
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	var err error
+	interface_, err = newTestInterface()
+	if err != nil {
+		fmt.Printf("Error creating interface: %v", err)
+		os.Exit(1)
+	}
+	os.Exit(m.Run())
+}
+
+// firstZone returns the first zone for the configured dns provider account/project,
+// or fails if it can't be found
+func firstZone(t *testing.T) dnsprovider.Zone {
+	i := interface_
+	t.Logf("Getting zones")
+	z, supported := i.Zones()
+	if supported {
+		t.Logf("Got zones %v\n", z)
+	} else {
+		t.Fatalf("Zones interface not supported by interface %v", i)
+	}
+	zones, err := z.List()
+	if err != nil {
+		t.Fatalf("Failed to list zones: %v", err)
+	} else {
+		t.Logf("Got zone list: %v\n", zones)
+	}
+	if len(zones) < 1 {
+		t.Fatalf("Zone listing returned %d, expected %d", len(zones), 1)
+	} else {
+		t.Logf("Got at least 1 zone in list:%v\n", zones[0])
+	}
+	return zones[0]
+}
+
+/* rrs returns the ResourceRecordSets interface for a given zone */
+func rrs(t *testing.T, zone dnsprovider.Zone) (r dnsprovider.ResourceRecordSets) {
+	rrsets, supported := zone.ResourceRecordSets()
+	if !supported {
+		t.Fatalf("ResourceRecordSets interface not supported by zone %v", zone)
+		return r
+	}
+	return rrsets
+}
+
+func listRrsOrFail(t *testing.T, rrsets dnsprovider.ResourceRecordSets) []dnsprovider.ResourceRecordSet {
+	rrset, err := rrsets.List()
+	if err != nil {
+		t.Fatalf("Failed to list recordsets: %v", err)
+	} else {
+		if len(rrset) < 0 {
+			t.Fatalf("Record set length=%d, expected >=0", len(rrset))
+		} else {
+			t.Logf("Got %d recordsets: %v", len(rrset), rrset)
+		}
+	}
+	return rrset
+}
+
+func getExampleRrs(zone dnsprovider.Zone) dnsprovider.ResourceRecordSet {
+	rrsets, _ := zone.ResourceRecordSets()
+	return rrsets.New("www11."+zone.Name(), []string{"10.10.10.10", "169.20.20.20"}, 180, rrstype.A)
+}
+
+func getInvalidRrs(zone dnsprovider.Zone) dnsprovider.ResourceRecordSet {
+	rrsets, _ := zone.ResourceRecordSets()
+	return rrsets.New("www12."+zone.Name(), []string{"rubbish", "rubbish"}, 180, rrstype.A)
+}
+
+func addRrsetOrFail(t *testing.T, rrsets dnsprovider.ResourceRecordSets, rrset dnsprovider.ResourceRecordSet) dnsprovider.ResourceRecordSet {
+	result, err := rrsets.Add(rrset)
+	if err != nil {
+		t.Fatalf("Failed to add recordsets: %v", err)
+	}
+	return result
+}
+
+/* TestResourceRecordSetsList verifies that listing of zones succeeds */
+func TestZonesList(t *testing.T) {
+	firstZone(t)
```

Comment:
This is quite non-idiomatic. Usually Go tests write their errors and let other tests to proceed. That way, you can see all the failing tests and the reason for their failures. So in the next cycle you can fix them together. Crashing on the first failure prevents that from happening. 

It is same argument as compiler/parser errors. Should the parser fail upon seeing the first error or continue to do its best to report errors until it can't? Experience has taught us how frustrating the latter could be :)